### DISCUSSION TITLE
Expose an `AxisItemPlacer.Horizontal` implementation that allows manual filtering of values

### DISCUSSION THREAD

**Discussion by kevincianfarini** [patrykandpatrick/vico#621] [Ideas]
I have a requirement in our graphing features to display X-axis labels for the times of the day. Specifically, I need to expose: 

* 00:00
* 06:00
* 12:00
* 18:00
* 23:00

The interval spacing of these axis labels are not uniform, and therefore I can't use the default horizontal placer with a spacing of 1. 

I'm thinking a dynamic API for this kind of thing would be very nice. Perhaps something like the following? 

```kt
// Companion object in AxisItemPlacer.Horizontal
public companion object {
  
  fun dynamic(predicate: (xIndex: Float) -> Boolean) = DynamicHorizontalItemPlacer(...)
}

// usage
itemPlacer = AxisItemPlacer.Horizontal.dynamic { xIndex: Float
  xIndex.toInt() % 6 == 0 || xIndex.toInt() == 23
}
```

If there's an easier way to achieve this behavior without providing a fully custom implementation of `AxisItemPlacer.Horizontal` please let me know! I wasn't able to find a way to do so.

**Comment by patrickmichalik** [comment]
Hello! Thanks for the suggestion. This would certainly be a useful addition—we’ll look into it.

For the time being, to implement the behavior in question, you’ll have to create a custom `AxisItemPlacer.Horizontal` implementation. You could write it from scratch or use [the `DefaultHorizontalAxisItemPlacer` definition](https://github.com/patrykandpatrick/vico/blob/v2.0.0-alpha.12/vico/core/src/main/java/com/patrykandpatrick/vico/core/axis/horizontal/DefaultHorizontalAxisItemPlacer.kt#L29) as a starting point, copying it over and adding the required check.

**Reply by kevincianfarini** [reply]
Neat, thanks. If I get an implementation working well would you accept a PR?

**Reply by patrickmichalik** [reply]
Certainly, @kevincianfarini! Thanks for your interest in contributing. My suggestion would be to make this a feature of `DefaultHorizontalAxisItemPlacer`, but we’re open to other ideas. Let me know if you have any questions.

**Reply by kevincianfarini** [reply]
@patrickmichalik I have a rudimentary implementation of dynamic horizontal labels working. 

```kt
override fun getLabelValues(
    context: ChartDrawContext,
    visibleXRange: ClosedFloatingPointRange<Float>,
    fullXRange: ClosedFloatingPointRange<Float>,
    maxLabelWidth: Float,
): List<Float> = buildList {
    val start = maxOf(visibleXRange.start, context.chartValues.minX)
    val endInclusive = minOf(visibleXRange.endInclusive, context.chartValues.maxX)
    var currentIndex = start
    while (currentIndex <= endInclusive) {
        if (labelSelector(currentIndex)) add(currentIndex)
        currentIndex += context.chartValues.xStep
    }
}
```

And it's usage looks like this: 

```kt
itemPlacer = remember(xAxisLabelSelector, values) {
    DynamicHorizontalItemPlacer { xIndex ->
        xAxisLabelSelector(values[xIndex.toInt()])
    }
},

```

One thing I'm struggling with is that the first index's and last index's labels are being truncated, presumably because they don't have enough space to fit in their allocated area. Do you have any idea how to stop this truncation without also expanding the range of values in the X domain? In the below screenshot, I am attempting to display labels for `00:00`, `06:00`, `12:00`, `18:00`, and `23:00`. As you can see the labels for `00:00` and `23:00` get ellipsized. 

![Screenshot_1711479886](https://github.com/patrykandpatrick/vico/assets/13242326/60c282af-f18f-481f-a7c4-188e4ae29a51)

**Reply by patrickmichalik** [reply]
@kevincianfarini, overriding `getFirstLabelValue` and `getLastLabelValue` to return the appropriate _x_ values should resolve the truncation issue. You can find an example of this in `DefaultHorizontalAxisItemPlacer`. Please take note of the `addExtremeLabelPadding` property.

In the example you’ve shared, it looks like a list previously used for a `Transaction` is directly referenced in the lambda. However, there’s no guarantee that the `Transaction` has already been processed—`values` may not match what’s shown on the chart. My suggestion would be to retrieve the current chart data in `getLabelValues` via `context.chartValues` and pass it to the lambda. `ChartValues#model` will most likely be of interest. Extras can be used if needed.

If you’re looking to open a PR, here’s some more feedback:

1. Please note that the manner in which `start` and `endInclusive` are calculated here is incompatible with scrollable charts—`maxOf(visibleXRange.start, context.chartValues.minX)`, for example, won’t necessarily produce a multiple of the _x_ step.
2. Please keep in mind that while a chart’s _x_ values can be used as indices, they don’t have to be. Thus, for built-in APIs, no such assumption should be made, and names like `xIndex` should be avoided. We generally use `x`.
3. It would be worth considering whether calculations based on the chart data should be performed in the lambda. The lambda could instead have an `ExtraStore` parameter, in which case the user would precalculate the label _x_ values and save them as an extra. I’d love to hear your thoughts on this. Would such an approach be sufficient for your use case? `AxisItemPlacer.Vertical.step` and `AxisItemPlacer.Vertical.count`, for example, have `ExtraStore` lambdas.
4. I believe a lot of `DefaultHorizontalAxisItemPlacer` logic will have to be duplicated here, which is why I’d consider updating `DefaultHorizontalAxisItemPlacer` rather than introducing a new `AxisItemPlacer.Horizontal` implementation. I believe a filtering lambda could be added to `DefaultHorizontalAxisItemPlacer` relatively easily.

Thanks!

**Reply by kevincianfarini** [reply]
> * Please note that the manner in which start and endInclusive are calculated here is incompatible with scrollable charts—maxOf(visibleXRange.start, context.chartValues.minX), for example, won’t necessarily produce a multiple of the x step.
>* Please keep in mind that while a chart’s x values can be used as indices, they don’t have to be. Thus, for built-in APIs, no such assumption should be made, and names like xIndex should be avoided. We generally use x.
> * I believe a lot of DefaultHorizontalAxisItemPlacer logic will have to be duplicated here, which is why I’d consider updating DefaultHorizontalAxisItemPlacer rather than introducing a new AxisItemPlacer.Horizontal implementation. I believe a filtering lambda could be added to DefaultHorizontalAxisItemPlacer relatively easily.

I've decided to simplify this a bit and lean more heavily on the default implementation. The following implementation could certainly be tweaked to be a feature of the default implementation, but I'm opting for composition here so that I can iterate quickly. It looks like this now: 

```kt
internal class DynamicHorizontalItemPlacer(
    private val delegate: AxisItemPlacer.Horizontal = AxisItemPlacer.Horizontal.default(),
    private val xValueSelector: (Float) -> Boolean,
) : AxisItemPlacer.Horizontal by delegate {

    override fun getLabelValues(
        context: ChartDrawContext,
        visibleXRange: ClosedFloatingPointRange<Float>,
        fullXRange: ClosedFloatingPointRange<Float>,
        maxLabelWidth: Float
    ): List<Float> = delegate.getLabelValues(context, visibleXRange, fullXRange, maxLabelWidth)
        .filter(xValueSelector)
}
```

> It would be worth considering whether calculations based on the chart data should be performed in the lambda. The lambda could instead have an ExtraStore parameter, in which case the user would precalculate the label x values and save them as an extra. I’d love to hear your thoughts on this. Would such an approach be sufficient for your use case? AxisItemPlacer.Vertical.step and AxisItemPlacer.Vertical.count, for example, have ExtraStore lambdas.

I hadn't considered this. I'm still evaluating if Vico fits our use cases so I don't have much experience using it. I can certainly look into how this might work in practice. Since this is all implementation details, is there any detriment to having the lambda filter parameter as an instance variable versus storing it in `ExtraStore`? The public API would be the same regardless, right?

**Reply by kevincianfarini** [reply]
@patrickmichalik Specifying `addExtremeLabelPadding = true` unfortunately doesn't solve this for me. It adds gaps on either side of the X domain and for some reason only renders one X-axis label. I'm not sure why. Am I perhaps bumping up against an inherent limitation of the axis item placer? If so I can talk to my designer -- I wouldn't think this would be a deal breaker for us. 

![Screenshot_1711551177](https://github.com/patrykandpatrick/vico/assets/13242326/2a0b70a7-3242-47aa-904d-ad9ddc024749)

I'm wondering if there's any way I can shift labels towards the end or start if they're not going to fit on the axis when they're measured without truncating. Do you think that's possible?

**Reply by kevincianfarini** [reply]
> In the example you’ve shared, it looks like a list previously used for a Transaction is directly referenced in the lambda. However, there’s no guarantee that the Transaction has already been processed—values may not match what’s shown on the chart. My suggestion would be to retrieve the current chart data in getLabelValues via context.chartValues and pass it to the lambda. ChartValues#model will most likely be of interest. Extras can be used if needed.

I see what you mean now. I will update this. Thank you!

**Reply by patrickmichalik** [reply]
Using `AxisItemPlacer.Horizontal.default` like this might be problematic, @kevincianfarini, as `DefaultHorizontalAxisItemPlacer` has automatic features that may get in the way here. This wouldn’t be an issue if `DefaultHorizontalAxisItemPlacer` were to be modified, as you could then update the existing logic to prevent these features from being enabled when they’re not desired. For your project, I’d suggest either copying over and modifying the `DefaultHorizontalAxisItemPlacer` definition or using a fully custom `AxisItemPlacer.Horizontal` implementation. I’ve taken the latter approach and created the following chart:

<img src="https://github.com/patrykandpatrick/vico/assets/120058021/a7d47bc7-f278-4f80-890b-b197290fbd7f" width="720" />

We don’t yet have a built-in API for shifting a `HorizontalAxis`’s extreme labels inward, but we plan on adding such a feature. You can, of course, create a custom `Axis` implementation with such functionality. Here, I’ve copied the `HorizontalAxis` definition and overridden the label positioning logic for the first and last labels so that they’re drawn as you’ve described:

<img src="https://github.com/patrykandpatrick/vico/assets/120058021/550c893b-409b-480e-8f1c-6dbffe3092e8" width="720" />

Let me know if you’d like me to clean up and share the code for either one of these charts.

Regarding the `ExtraStore` idea, the lambda parameters would differ—rather than receiving the chart data, the lambda would receive the `ExtraStore`. The label _x_ values would be determined at the `Transaction` stage, which would reduce the number of operations performed on the main thread during drawing. In the case of your chart, the difference between these two solutions is negligible—I brought this up only in case you wish to open a pull request down the line.

**Reply by kevincianfarini** [reply]
If you could share some code that would be wonderful. My current objective is to get Vico adopted on my team, and then potentially help upstream some work we do.

**Reply by patrickmichalik** [reply]
Sounds good, @kevincianfarini. Just to confirm, would you be interested in the code for the first chart or for the second one? The first chart is easier to implement, if that’s relevant.

I’d also be happy to share the code for both charts, but if you’d like me to do so, please let me know which one to start with. I may need some time to clean everything up, so it’d be helpful to know what to prioritize. Thanks!

**Reply by kevincianfarini** [reply]
> Just to confirm, would you be interested in the code for the first chart or for the second one? The first chart is easier to implement, if that’s relevant.

The second chart is what I'm trying to implement, so sharing that would be especially helpful 😄

**Reply by kevincianfarini** [reply]
Following up on this. I've gotten our usage of Vico to a great place thanks to all of your help and work @patrickmichalik! Thank you so much. 

The last thing I'm waiting on before I can merge is ensuring X-axis labels aren't truncated. Whenever you have time I'd love to see your code sample from above 😄

**Reply by patrickmichalik** [reply]
I’m glad to hear that, @kevincianfarini. You’re welcome! I’ll prepare the code sample for you as soon as I can.

**Reply by kevincianfarini** [reply]
Woohoo! Thank you 🙌

**Reply by kevincianfarini** [reply]
Ping :smile:

**Reply by kevincianfarini** [reply]
Hi @patrickmichalik! We've merged our Vico implementation without these offset labels. Is there anything I can help do in order to get the sample code from the above screenshot?

**Reply by patrickmichalik** [reply]
Apologies for the wait, @kevincianfarini. I’ve been quite busy due to Vico updates and other matters. Please rest assured that I’ll look into this when possible. I should be able to do so this week.

**Reply by kevincianfarini** [reply]
Thank you!

**Reply by patrickmichalik** [reply]
The example is almost ready, @kevincianfarini. I had some unforseen things come up, which led to a slight delay, but I expect to finalize and share the code tomorrow. If Saturday is the last day of the week for you, this will be a bit later than originally predicted, hence the update. Thanks for your patience.

**Reply by kevincianfarini** [reply]
No worries! Thank you so much. There's no rush, I just wanted to make sure I didn't slip through the cracks :smile:

**Reply by patrickmichalik** [reply]
Please find the code [here](https://gist.github.com/patrickmichalik/b2fc0a480d901f4c072357a74fbed7ce), @kevincianfarini. I’ve made some modifications for better alignment with your chart and with the snippets that you’ve shared. A few comments:

1. This is compatible with Vico 2.0.0 Alpha 15 and later. Notably, the ability to create `HorizontalAxis` and `VerticalAxis` subclasses was added in Vico 2.0.0 Alpha 15. Previously, one had to copy the definitions.
2. `BottomAxis` handles only those `HorizontalAxis` features that your chart uses. `BaseAxis#guideline`, for example, is ignored.
3. The assumption is made that all labels will fit. I didn’t add any logic for overlap prevention, as I’m not familiar with what your requirements are in that respect.
4. `labelPredicate` has a `CartesianDrawContext` receiver type. This isn’t necessary for the example—I added it in case you need to access the `CartesianChartModel` from the lambda. This relates to the second paragraph [here](https://github.com/patrykandpatrick/vico/discussions/621#discussioncomment-8920837).

If you need any help adapting this for an older Vico version (point 1), adding support for an unhandled `HorizontalAxis` feature (point 2), or implementing overlap prevention (point 3), please let me know. I’ll also be happy to answer any other questions you may have.

<img src="https://github.com/patrykandpatrick/vico/assets/120058021/6c245789-cbd3-4d1e-b260-a652dd1e5af4" width="720" />

**Reply by kevincianfarini** [reply]
Thank you so much!

**Reply by patrickmichalik** [reply]
You’re welcome!
